mr/src/main/java/org/apache/iceberg/mr/mapred/iterables/SnapshotIterable.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.mapred.iterables;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+
+/**
+ * Creates an Iterable of Records with all snapshot metadata that can be used with the RecordReader.
+ */
+public class SnapshotIterable implements CloseableIterable {
```

Comment:
Is this really needed? Can we reuse https://github.com/apache/iceberg/blob/cad124967ceb740178a3f65b75f19017494b430d/core/src/main/java/org/apache/iceberg/SnapshotsTable.java#L60 to get the iterable instead?

Reply:
The reason we ended up adding the `SnapshotIterable` is so we could wrap a row from the SnapshotsTable in a `Record` which could be passed back to the `RecordReader` without needing to change any code for the RR. It could be super helpful if there was a method in `SnapshotsTable` that returned Records for the whole table, as there isn't a Reader, so to speak, for the `METADATA` file format (https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/FileFormat.java#L31) that we can add in to the `ReaderFactory` that would allow us 'read' from the SnapshotTable

Reply:
My concern is that we will need N such classes for the N metadata tables we offer. I would suggest looking at how this is handled for Spark.

https://github.com/apache/iceberg/blob/cad124967ceb740178a3f65b75f19017494b430d/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java#L108

Reply:
Also we can probably consider adding support for metadata tables in a separate PR? This PR has a lot of new functionality to review.

Reply:
A great call out, that looks like exactly the sort of thing we should be doing here! I'll remove most of the system tables stuff from our PR's and we'll do a separate PR that should be easier to review :))